### PR TITLE
Allow setting candidates for a request

### PR DESCRIPTION
**Pull request by loldot**
- [x] New feature
- [ ] Bug fix
- [ ] High impact

**Description of work:**
AB#28176
Allow resource owners to make a shortlist of candidates for a request before proposing.

**Testing:**
- [x] Can be tested
- [x] Automatic tests created / updated
- [x] Local tests are passing

**Checklist:**
- [x] Considered automated tests
- [x] Considered updating specification / documentation
- [x] Considered work items 
- [x] Considered security
- [x] Performed developer testing
- [x] Checklist finalized / ready for review

### CODE REVIEW

File: src/backend/api/Fusion.Resources.Api/Controllers/Requests/InternalRequestsController.cs [hunk 1]
```diff
@@ -421,6 +421,11 @@ public InternalRequestsController(IFusionProfileResolver profileResolver)
                     updateCommand.ProposalScope = @params.ResolveScope();
                     updateCommand.ProposalChangeType = @params.Type;
                 }
+                
+                if (request.Candidates.HasValue)
+                {
+                    updateCommand.Candidates = request.Candidates.Value.Select(x => (PersonId)x).ToList();
```
Comment (xssic) [inline src/backend/api/Fusion.Resources.Api/Controllers/Requests/InternalRequestsController.cs:427]: possible null ref

File: src/backend/api/Fusion.Resources.Database/Entities/DbPerson.cs [hunk 2]
```diff
@@ -18,6 +19,7 @@ public class DbPerson
 
         public string AccountType { get; set; } = null!;
         public string? JobTitle { get; set; }
+        public List<DbResourceAllocationRequest> CandidateFor { get; set; } = new();
```
Comment (xssic) [inline src/backend/api/Fusion.Resources.Database/Entities/DbPerson.cs:22]: Candidate for what?
Maybe a better specified name ?

File: src/backend/api/Fusion.Resources.Database/Entities/DbResourceAllocationRequest.cs [hunk 3]
```diff
@@ -57,7 +57,16 @@ public class DbResourceAllocationRequest
         public DbOpProposedPerson ProposedPerson { get; set; } = DbOpProposedPerson.Empty;
         public DbOpProposalParameters ProposalParameters { get; set; } = new DbOpProposalParameters();
 
+        public void ProposePerson(DbPerson person, DbOpProposalParameters? parameters = null)
```
Comment (xssic) [inline src/backend/api/Fusion.Resources.Database/Entities/DbResourceAllocationRequest.cs:60]: nullable or not. mismatch between argument and usage below

File: src/backend/api/Fusion.Resources.Database/Entities/DbResourceAllocationRequest.cs [hunk 4]
```diff
@@ -69,10 +78,11 @@ public class DbResourceAllocationRequest
         public DateTimeOffset LastActivity { get; set; }
 
         public DbOpProvisionStatus ProvisioningStatus { get; set; } = new DbOpProvisionStatus();
-        public List<DbRequestAction>? Actions { get;  set; }
+        public List<DbRequestAction>? Actions { get; set; }
         public List<DbConversationMessage>? Conversation { get; set; }
 
         public Guid? CorrelationId { get; set; }
+        public List<DbPerson> Candidates { get; set; }
```
Comment (xssic) [inline src/backend/api/Fusion.Resources.Database/Entities/DbResourceAllocationRequest.cs:85]: shoul be non-null when exiting constructor (build warning)

File: src/backend/api/Fusion.Resources.Domain/Commands/Requests/InternalRequests/UpdateInternalRequest.cs [hunk 5]
```diff
@@ -50,7 +50,7 @@ public Handler(ResourcesDbContext db, IProfileService profileService, IMediator
 
             public async Task<QueryResourceAllocationRequest> Handle(UpdateInternalRequest request, CancellationToken cancellationToken)
             {
-                var dbRequest = await db.ResourceAllocationRequests.FirstAsync(r => r.Id == request.RequestId);
+                var dbRequest = await db.ResourceAllocationRequests.FirstAsync(r => r.Id == request.RequestId, cancellationToken);
```
Comment (xssic) [inline src/backend/api/Fusion.Resources.Domain/Commands/Requests/InternalRequests/UpdateInternalRequest.cs:53]: During modification of data/state. Do we really want to be able to halt execution in the middle of whatever happens ?

File: src/backend/tests/Fusion.Resources.Api.Tests/IntegrationTests/InternalResourceAllocationRequestTests.cs [hunk 6]
```diff
@@ -947,6 +948,64 @@ public async Task UpdateRequest_ProposedPersonShouldNotBeNullable_WhenApproving(
             );
             response.Should().BeBadRequest();
         }
+
```
Comment (xssic) [inline src/backend/tests/Fusion.Resources.Api.Tests/IntegrationTests/InternalResourceAllocationRequestTests.cs:951]: whenever possible. add both negative and positive tests

Comment (loldot) [reply src/backend/tests/Fusion.Resources.Api.Tests/IntegrationTests/InternalResourceAllocationRequestTests.cs:951]: Positive cases should already be covered by existing tests.

**Review by xssic — APPROVED** [review]

**Pull request merged**